paasta_tools/autoscaling_lib.py

```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
+    """Scales a spot fleet request by delta to reach target capacity
+    If scaling up we just set target capacity and let AWS take care of the rest
+    If scaling down we pick the slaves we'd prefer to kill, put them in maintenance
+    mode and drain them (via paasta_maintenance and setup_marathon_jobs). We then kill
+    them once they are running 0 tasks or once a timeout is reached
+
+    :param resource: resource to scale
+    :param delta: integer change in number of servers
+    :param target_capacity: target number of instances
+    :param sorted_slaves: list of slaves by order to kill
+    :param dry_run: Don't drain or make changes to spot fleet if True"""
+    sfr_id = resource['id']
+    ec2_client = boto3.client('ec2')
+    if delta == 0:
+        return
+    elif delta > 0:
+        log.debug("Increasing spot fleet capacity to: {0}".format(target_capacity))
+        if not dry_run:
+            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
+                                                 ExcessCapacityTerminationPolicy='noTermination')
+            return
+    elif delta < 0:
+        number_to_kill = delta * -1
+        sfr_ips = get_sfr_instance_ips(sfr_id)
+        log.debug("IPs in SFR: {0}".format(sfr_ips))
+        sfr_sorted_slaves = [slave for slave in sorted_slaves if slave_pid_to_ip(slave['pid']) in sfr_ips]
+        log.debug("SFR slave kill preference: {0}".format(sfr_sorted_slaves))
+        if number_to_kill > len(sfr_sorted_slaves):
+            log.error("Didn't find enough candidates to kill. This shouldn't happen so let's not kill anything!")
+            return
+        else:
+            slaves_to_kill = sfr_sorted_slaves[0:number_to_kill]
+        log.debug("Set to kill: {0}".format([slave['pid'] for slave in slaves_to_kill]))
+        instances_to_kill = {}
+        for slave in slaves_to_kill:
+            ip = slave_pid_to_ip(slave['pid'])
+            instances_to_kill[slave['pid']] = {'ip': ip}
+            instances_to_kill[slave['pid']]['instance_id'] = get_instance_id_from_ip(ip)
+        # The start time of the maintenance window is the point at which
+        # we giveup waiting for the instance to drain and mark it for termination anyway
+        start = int(time.time() + CLUSTER_DRAIN_TIMEOUT)
+        # Set the duration to an hour, if we haven't cleaned up and termintated by then
+        # mesos should put the slave back into the pool
+        duration = 600
+        log.debug("Draining {0}".format(instances_to_kill))
+        log.debug("Decreasing spot fleet capacity to: {0}".format(target_capacity))
+        if not dry_run:
+            # sort to make testing easier
+            drain([instance['ip'] for instance in sorted(instances_to_kill.values())], start, duration)
+            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
```

Comment:
I'm a little concerned about the method here.

What happens if something goes wrong with the `wait_and_terminate` step? I don't think the autoscaler will be able to recover because those instances will be "lost".

I think you have to do it "atomically" where scale + kill at the same time on each instance, otherwise you risk instances falling through the cracks